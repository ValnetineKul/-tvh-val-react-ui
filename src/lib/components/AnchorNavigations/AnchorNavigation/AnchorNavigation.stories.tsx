import type { ComponentProps } from 'react';
import React from 'react';
import { createMeta, createTemplate } from '../../../story-utils';
import Typography from '../../Typography';
import TabNavigation from '../../TabNavigation';
import AnchorNavigation from './AnchorNavigation';

type Props = ComponentProps<typeof AnchorNavigation>;

const tabs = [
  {
    id: 'T1',
    label: 'First tab',
    component: 'First tab content',
  },
  {
    id: 'T2',
    label: 'Second tab',
    component: 'Second tab content',
  },
  {
    id: 'T3',
    label: 'Third tab',
    component: 'Third tab content',
  },
];

const items: Props['items'] = [
  {
    label: 'Will it fit',
    component: (
      <div style={{ paddingTop: '16px' }}>
        <TabNavigation tabList={tabs} />
      </div>
    ),
    surface: '200',
  },
  {
    label: 'About the product',
    component: (
      <div style={{ height: '20vh' }}>
        <Typography variant="body500">About the product anchor content.</Typography>
      </div>
    ),
    surface: '100',
  },
  {
    label: 'Alternatives',
    component: (
      <div style={{ height: '20vh' }}>
        <Typography variant="body500">Alternatives anchor content.</Typography>
      </div>
    ),
    surface: '300',
  },
  {
    label: 'Related items',
    component: (
      <div style={{ height: '20vh' }}>
        <Typography variant="body500">Related items anchor content.</Typography>
      </div>
    ),
  },
  {
    label: 'Composition',
    component: (
      <div style={{ height: '20vh' }}>
        <Typography variant="body500">Composition anchor content.</Typography>
      </div>
    ),
    surface: '200',
  },
  {
    label: 'Frequently bought together',
    component: (
      <div style={{ height: '20vh' }}>
        <Typography variant="body500">Frequently bought together anchor content.</Typography>
      </div>
    ),
  },
  {
    label: 'Extra long AnchorItem label which is longer than 30 characters',
    component: (
      <div style={{ height: '20vh' }}>
        <Typography variant="body500">
          The maximum amount of characters for each AnchorItem is 30 (including spaces). When the label is longer than
          that, it will be truncated &amp; a Tooltip will be shown on hover.
        </Typography>
      </div>
    ),
    surface: '200',
  },
];

const Template = createTemplate(AnchorNavigation);

export const Primitive = Template.bind({});

Primitive.args = {
  items,
};

Primitive.storyName = 'AnchorNavigation';

export default createMeta({
  component: AnchorNavigation,
  title: 'Components/AnchorNavigations/AnchorNavigation',
  argTypes: {
    items: {
      control: {
        disable: true,
      },
    },
  },
});
