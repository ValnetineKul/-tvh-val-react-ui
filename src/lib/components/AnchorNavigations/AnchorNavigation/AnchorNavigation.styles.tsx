import { makeStyles } from '../../../../themes/core';

const useStyles = makeStyles<{ offset: number }>({ name: 'AnchorNavigation' })((theme, { offset }) => ({
  root: {
    position: 'sticky',
    top: `${offset}px`,
    zIndex: 10,
  },
  anchorList: {
    minHeight: 'auto',
    width: '100%',
  },
  indicator: {
    display: 'none',
  },
  iconButton: {
    padding: 0,
    '&:focus': {
      outline: 'none',
    },
  },
  anchorComponent: {
    paddingTop: theme.layoutSpacing['spacing/450'],
    paddingBottom: theme.layoutSpacing['spacing/450'],
  },
  header: {
    marginBottom: theme.layoutSpacing['spacing/400'],
  },
}));

export default useStyles;
