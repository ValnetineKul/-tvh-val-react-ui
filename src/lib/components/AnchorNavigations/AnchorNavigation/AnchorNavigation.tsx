import type { FC, ReactNode } from 'react';
import React from 'react';
import Tabs from '@mui/material/Tabs';
import IconButton from '../../Buttons/IconButton';
import Container from '../../Container';
import { AngleLeft, AngleRight } from '../../Icon/icons/functional';
import Icon from '../../Icon';
import Surface from '../../Surfaces/Surface';
import Typography from '../../Typography';
import AnchorItem from '../AnchorItem';
import useStyles from './AnchorNavigation.styles';
import type { DataAttributes } from '../../../types/common';

interface AnchorNavigationItem {
  label: string;
  component: ReactNode;
  surface?: '100' | '200' | '300';
  anchorLinkProps?: React.HTMLAttributes<HTMLAnchorElement> & DataAttributes;
}

export interface AnchorNavigationProps {
  items: AnchorNavigationItem[];
  headersType?: 'functional' | 'commercial';
  offset?: number;
}

const AnchorNavigation: FC<AnchorNavigationProps> = ({ items, headersType = 'functional', offset = 0 }) => {
  const { classes } = useStyles({ offset });

  const ScrollButtonComponent = (props: React.ComponentProps<typeof IconButton> & { direction: 'left' | 'right' }) => {
    if (props.direction === 'left') {
      return <IconButton {...props} className={classes.iconButton} focus="inset" icon={<Icon icon={AngleLeft} />} />;
    }
    if (props.direction === 'right') {
      return <IconButton {...props} className={classes.iconButton} focus="inset" icon={<Icon icon={AngleRight} />} />;
    }
    return null;
  };

  const createId = (key: number) => `anchor-navigation-${key}`;
  const itemsWithIds = items.map((item, key) => ({ ...item, id: createId(key) }));

  const renderAnchorItems = (item: AnchorNavigationItem & { id: string }, key: number) => {
    return (
      // workaround with offset because active class is not set on initial render in react-scroll https://github.com/fisshy/react-scroll/issues/104
      <AnchorItem
        item={item}
        offset={key === 0 ? -(offset + 60) : -(offset + 40)}
        key={key}
        anchorLinkProps={item.anchorLinkProps}
      />
    );
  };

  const renderContent = (item: AnchorNavigationItem, key: number) => {
    const { label, component, surface = '100' } = item;

    return (
      <section id={createId(key)} data-navitem={label} key={key}>
        <Surface color={surface} className={classes.anchorComponent}>
          <Container>
            <Typography variant="h2" headerType={headersType} className={classes.header}>
              {label}
            </Typography>
            {component}
          </Container>
        </Surface>
      </section>
    );
  };

  return (
    <>
      <Surface color="300" className={classes.root} component="nav">
        <Container>
          <Tabs
            value={false}
            variant="scrollable"
            scrollButtons="auto"
            ScrollButtonComponent={ScrollButtonComponent}
            classes={{
              root: classes.anchorList,
              indicator: classes.indicator,
            }}
          >
            {itemsWithIds.map(renderAnchorItems)}
          </Tabs>
        </Container>
      </Surface>
      {items.map(renderContent)}
    </>
  );
};

export default AnchorNavigation;
