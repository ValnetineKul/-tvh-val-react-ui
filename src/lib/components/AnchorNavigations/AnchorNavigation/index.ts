import AnchorNavigation from './AnchorNavigation';

export type { AnchorNavigationProps } from './AnchorNavigation';
export default AnchorNavigation;
