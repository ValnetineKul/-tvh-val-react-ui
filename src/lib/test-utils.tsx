import userEvent from '@testing-library/user-event';
import { fireEvent, render } from '@testing-library/react';
import { renderHook } from '@testing-library/react-hooks';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { theme } from '../themes/tvh';
import { createTheme, ThemeProvider, SurfaceProvider } from '../themes/core';

export const muiTheme = createTheme(theme);

const AllTheProviders: React.FC = ({ children }) => {
  return (
    <MemoryRouter>
      <ThemeProvider theme={muiTheme}>
        <SurfaceProvider color="100">{children}</SurfaceProvider>
      </ThemeProvider>
    </MemoryRouter>
  );
};

const customRender = (component: React.ReactElement, options?: Parameters<typeof render>[1]) =>
  render(component, { wrapper: AllTheProviders, ...options });

const customRenderHook: typeof renderHook = (hook, options) =>
  renderHook(hook, { wrapper: AllTheProviders, ...options });

export const initResponsiveTest = (size: number | 'Mobile' | 'Tablet' | 'Desktop' | 'LgDesktop' | 'XlDesktop') => {
  let width: number;
  if (size === 'Mobile') width = muiTheme.breakpoints.values.sm - 1;
  else if (size === 'Tablet') width = muiTheme.breakpoints.values.sm;
  else if (size === 'Desktop') width = muiTheme.breakpoints.values.md;
  else if (size === 'LgDesktop') width = muiTheme.breakpoints.values.lg;
  else if (size === 'XlDesktop') width = muiTheme.breakpoints.values.xl;
  else width = size;

  Object.defineProperty(window, 'innerWidth', {
    writable: true,
    configurable: true,
    value: width,
  });

  window.matchMedia = jest.fn().mockImplementation((query) => {
    const queryWidthInterval: string[] = query.match(/\d+(\.\d+)?/g);

    let matches;
    if (queryWidthInterval.length === 2) {
      matches = width >= +queryWidthInterval[0] && width < +queryWidthInterval[1];
    } else {
      const queryWidth = +queryWidthInterval[0];
      const isDown = /max/.test(query);

      matches = isDown ? width <= queryWidth : width >= queryWidth;
    }

    return {
      matches,
      media: query,
      onchange: null,
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  });
};

const customUserEvent = Object.assign(userEvent, {
  mouseDown: fireEvent.mouseDown,
  mouseUp: fireEvent.mouseUp,
  mouseMove: fireEvent.mouseMove,
  scroll: fireEvent.scroll,
  error: fireEvent.error,
});

export { screen, within, waitFor, act, waitForElementToBeRemoved } from '@testing-library/react';
// override render method
export { customRender as render, customRenderHook as renderHook, customUserEvent as userEvent };
