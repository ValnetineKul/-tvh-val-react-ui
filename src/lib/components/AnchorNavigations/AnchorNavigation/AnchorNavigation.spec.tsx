import React from 'react';
import { render } from '../../../test-utils';
import AnchorNavigation from './AnchorNavigation';

const items = [
  {
    label: 'link1',
    component: <div style={{ height: '20vh' }}>link1</div>,
  },
  {
    label: 'link2',
    component: <div style={{ height: '20vh' }}>link2</div>,
    surface: '100' as const,
  },
  {
    label: 'link3',
    component: <div style={{ height: '20vh' }}>link3</div>,
    surface: '200' as const,
  },
  {
    label: 'link4',
    component: <div style={{ height: '20vh' }}>link4</div>,
    surface: '300' as const,
  },
];

describe('AnchorNavigation', () => {
  it('Should render correctly', () => {
    const { container } = render(<AnchorNavigation items={items} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('Should not add paddingBottom', () => {
    const { getByRole, container } = render(<AnchorNavigation items={items} />);

    const wrapper = container.firstChild as HTMLDivElement;
    const lastComponent = wrapper.childNodes[wrapper.childNodes.length - 1];
    const nav = getByRole('navigation');

    Object.defineProperties(wrapper, {
      offsetHeight: {
        configurable: true,
        value: 1013,
      },
      offsetTop: {
        configurable: true,
        value: 954,
      },
    });

    Object.defineProperty(lastComponent, 'offsetHeight', { configurable: true, value: 760 });
    Object.defineProperty(document.body, 'offsetHeight', { configurable: true, value: 2606 });
    Object.defineProperty(nav, 'offsetHeight', { configurable: true, value: 58 });

    const paddingBottomValue = getComputedStyle(wrapper, null).getPropertyValue('padding-bottom');
    expect(paddingBottomValue).toBe('');
  });

  describe('Should set headers type', () => {
    it.each([
      [undefined, 'headerFunctional'],
      ['functional' as const, 'headerFunctional'],
      ['commercial' as const, 'headerCommercial'],
    ])('type = %s', (type, className) => {
      const { getByRole } = render(<AnchorNavigation items={items} headersType={type} />);
      const header = getByRole('heading', { name: 'link1' });
      expect(header.className).toMatch(className);
    });
  });
});
