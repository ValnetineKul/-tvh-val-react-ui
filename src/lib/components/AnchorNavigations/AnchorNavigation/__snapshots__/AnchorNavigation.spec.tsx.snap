// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnchorNavigation Should render correctly 1`] = `
<nav
  class="tss-11kuxjf-Surface-root-AnchorNavigation-root"
>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg tss-1nurcy4-Container-root mui-1oqqzyl-MuiContainer-root"
  >
    <div
      class="tss-1rb1cua-AnchorNavigation-anchorList MuiTabs-root mui-1ujnqem-MuiTabs-root"
    >
      <div
        class="MuiTabs-scrollableX MuiTabs-hideScrollbar mui-agdkjh-MuiTabs-scrollbarSize"
        style="width: 99px; height: 99px; position: absolute; top: -9999px; overflow: scroll;"
      />
      <div
        class="MuiTabs-scroller MuiTabs-hideScrollbar MuiTabs-scrollableX mui-69z67c-MuiTabs-scroller"
        style="margin-bottom: -100px;"
      >
        <div
          class="MuiTabs-flexContainer mui-heg063-MuiTabs-flexContainer"
          role="tablist"
        >
          <a
            class="MuiButtonBase-root MuiButtonBase-root MuiTab-root MuiTab-textColorInherit tss-1w7io4j-ButtonBase-root-ButtonBase-focusInset-AnchorItem-tabItem mui-1104c1u-MuiButtonBase-root-MuiButtonBase-root-MuiTab-root"
            href="#anchor-navigation-0"
            role="link"
            tabindex="0"
          >
            <div
              aria-label="link1"
              class="tss-2-Typography-weightRegular-ref tss-10drze0-Typography-root-Typography-weightRegular-Tooltip-wrapper-Typography-body400"
              data-mui-internal-clone-element="true"
            >
              <span
                class="tss-2-Typography-weightRegular-ref tss-78xlft-Typography-root-Typography-weightRegular-Typography-body500"
              >
                link1
              </span>
            </div>
          </a>
          <a
            class="MuiButtonBase-root MuiButtonBase-root MuiTab-root MuiTab-textColorInherit tss-1w7io4j-ButtonBase-root-ButtonBase-focusInset-AnchorItem-tabItem mui-1104c1u-MuiButtonBase-root-MuiButtonBase-root-MuiTab-root"
            href="#anchor-navigation-1"
            role="link"
            tabindex="0"
          >
            <div
              aria-label="link2"
              class="tss-2-Typography-weightRegular-ref tss-10drze0-Typography-root-Typography-weightRegular-Tooltip-wrapper-Typography-body400"
              data-mui-internal-clone-element="true"
            >
              <span
                class="tss-2-Typography-weightRegular-ref tss-78xlft-Typography-root-Typography-weightRegular-Typography-body500"
              >
                link2
              </span>
            </div>
          </a>
          <a
            class="MuiButtonBase-root MuiButtonBase-root MuiTab-root MuiTab-textColorInherit tss-1w7io4j-ButtonBase-root-ButtonBase-focusInset-AnchorItem-tabItem mui-1104c1u-MuiButtonBase-root-MuiButtonBase-root-MuiTab-root"
            href="#anchor-navigation-2"
            role="link"
            tabindex="0"
          >
            <div
              aria-label="link3"
              class="tss-2-Typography-weightRegular-ref tss-10drze0-Typography-root-Typography-weightRegular-Tooltip-wrapper-Typography-body400"
              data-mui-internal-clone-element="true"
            >
              <span
                class="tss-2-Typography-weightRegular-ref tss-78xlft-Typography-root-Typography-weightRegular-Typography-body500"
              >
                link3
              </span>
            </div>
          </a>
          <a
            class="MuiButtonBase-root MuiButtonBase-root MuiTab-root MuiTab-textColorInherit tss-1w7io4j-ButtonBase-root-ButtonBase-focusInset-AnchorItem-tabItem mui-1104c1u-MuiButtonBase-root-MuiButtonBase-root-MuiTab-root"
            href="#anchor-navigation-3"
            role="link"
            tabindex="0"
          >
            <div
              aria-label="link4"
              class="tss-2-Typography-weightRegular-ref tss-10drze0-Typography-root-Typography-weightRegular-Tooltip-wrapper-Typography-body400"
              data-mui-internal-clone-element="true"
            >
              <span
                class="tss-2-Typography-weightRegular-ref tss-78xlft-Typography-root-Typography-weightRegular-Typography-body500"
              >
                link4
              </span>
            </div>
          </a>
        </div>
        <span
          class="tss-2i5uv4-AnchorNavigation-indicator MuiTabs-indicator mui-1aquho2-MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
  </div>
</nav>
`;
